### PR TITLE
Support frondend in E2E transformer

### PR DESCRIPTION
**Pull request by kamo-naoyuki**
I modified E2E transformer to support frontend.

- [x] Add tests
- [x] Implement FrondendInterface and support dynamic importing scheme for fronend.

### CODE REVIEW

**Comment by sw005320** [comment]
Can you add a test script?

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 1]
```diff
@@ -224,7 +249,14 @@ def recognize(self, feat, recog_args, char_list=None, rnnlm=None, use_jit=False)
         TODO(karita): do not recompute previous attention for faster decoding
         '''
         self.eval()
-        feat = torch.as_tensor(feat).unsqueeze(0)
+        feat = to_torch_tensor(feat).unsqueeze(0)
+        # 0. Frontend
+        if self.frontend is not None:
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py]: I do not like this. Remember our `recognize` method is already too complex.

How about this?
```python

# espnet.nets.pytorch_backend.frontend_asr.py
class FrontendASR(torch.nn.Module):
    def __init__(self, frontend, asr_model):
        super().__init__()
        assert isinstance(asr_model, ASRInterface)
        self.frontend = frontend
        self.asr_model = asr_model

    def forward(self, xs_pad, ilens, ys_pad):
        hs_pad, hlens, _ = self.frontend(to_torch_tensor(xs_pad), ilens)
        hs_pad, hlens = self.feature_transform(hs_pad, hlens)
        return self.asr_model(hs_pad, hlens, ys_pad)

    def recognize(self, feat, recog_args, char_list=None, rnnlm=None, **kwargs):
        self.eval()
        feat = to_torch_tensor(feat).unsqueeze(0)
        feat, _, _ = self.frontend(feat, [feat.size(1)])
        feat, _ = self.feature_transform(feat, [feat.size(1)])
        feat = feat.squeeze(0)
        return self.asr_model(feat, recog_args, char_list, rnnlm, **kwargs)

    def enfance(...):
        ...
```

**Comment by kamo-naoyuki** [comment]
> Can you add a test script?

OK

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py (hunk 1 above)
Comment (ShigekiKarita) [reply espnet/nets/pytorch_backend/e2e_asr_transformer.py]: If `self.frontend.forward` applies `to_torch_tensor` inside itself, we can generalize this class between pytorch and chainer.

Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/e2e_asr_transformer.py]: I'm not sure what you propose. 
Does FrontendAsr correspond to the current `E2E`? I can't get  how to instantiate the class. Do you mean this?

```python
model = FrontendASR(FrondEnd(...), OriginalE2E(...))
```

Comment (ShigekiKarita) [reply espnet/nets/pytorch_backend/e2e_asr_transformer.py]: maybe you can add `--frontend-module espnet.nets.xxx:YYY` and asr.py can be changed like
``` python
model = dynamic_import(args.model_module)
...
if args.frontend_module is not None:
    frontend = dynamic_import(args.frontend_module)
    model = FrontendASR(frontend, model)
```

Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/e2e_asr_transformer.py]: O.K. @ShigekiKarita 's plan looks nice. 
I was also concerned about modularity of FrontEnd. 

I'll notify if I finished.

**Comment by kamo-naoyuki** [comment]
Finished. 
 I put out the fronted feature from E2E class and I implemented `FrontendASRInterface` and `FrontendASR`. Now we acquire two features.

- Frontend module can be used for arbitrary E2E class, including E2E for Transformer. 
- FrontendASR itself can be replaced by using `--frontend-module`. 

@Jzmo I changed `--use-frontend` to `--frontend-module`. Please reflect after merged.

File: espnet/nets/pytorch_backend/frontend_asr.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+from espnet.utils.cli_utils import strtobool
+
+
+class FrontendASR(FrontendASRInterface, torch.nn.Module):
+    """An implementation of FrontendASR with WPE and MVDR"""
+    @staticmethod
+    def add_arguments(parser: Union[argparse.ArgumentParser,
+                                    configargparse.ArgumentParser]):
+        # WPE related
+        parser.add_argument('--use-wpe', type=strtobool, default=False,
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/frontend_asr.py]: use group. see transformer

File: espnet/asr/pytorch_backend/asr.py [hunk 3]
```diff
@@ -609,11 +629,23 @@ def enhance(args):
...
+        # FIXME(kamo): Assuming fbank feature, so you need to change with the other frontend type
+        _idim = train_args.n_mels
+    else:
+        _idim = idim
+    model = model_class(_idim, odim, train_args)
+    assert isinstance(model, ASRInterface), type(model)
     model.recog_args = args
 
+    if train_args.frontend_module is None:
+        raise RuntimeError('--frontend_module is not set when training.')
```
Comment (ShigekiKarita) [inline espnet/asr/pytorch_backend/asr.py]: --frontend-module

File: espnet/asr/pytorch_backend/asr.py [hunk 4]
```diff
@@ -609,11 +629,23 @@ def enhance(args):
     # load trained model parameters
     logging.info('reading model parameters from ' + args.model)
     model_class = dynamic_import(train_args.model_module)
-    model = model_class(idim, odim, train_args)
-    assert isinstance(model, ASRInterface)
-    torch_load(args.model, model)
+    if train_args.frontend_module is not None:
+        # FIXME(kamo): Assuming fbank feature, so you need to change with the other frontend type
```
Comment (ShigekiKarita) [inline espnet/asr/pytorch_backend/asr.py]: remove this if block. FrontendASR can set its own `idim` by `train_args`

File: espnet/asr/pytorch_backend/asr.py [hunk 5]
```diff
@@ -247,9 +244,15 @@ def train(args):
 
     # specify model architecture
     model_class = dynamic_import(args.model_module)
-    model = model_class(idim, odim, args)
-    assert isinstance(model, ASRInterface)
+    if args.frontend_module is not None:
+        # FIXME(kamo): Assuming fbank feature, so you need to change with the other frontend type
+        _idim = args.n_mels
```
Comment (ShigekiKarita) [inline espnet/asr/pytorch_backend/asr.py]: ditto

File: espnet/nets/pytorch_backend/frontend_asr.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            uttmvn_norm_means=args.uttmvn_norm_means,
+            uttmvn_norm_vars=args.uttmvn_norm_vars)
+
+    def forward(self, xs_pad, ilens, ys_pad):
+        """E2E forward
+
+        :param torch.Tensor xs_pad: batch of padded input sequences (B, Tmax, idim)
+        :param torch.Tensor ilens: batch of lengths of input sequences (B)
+        :param torch.Tensor ys_pad: batch of padded character id sequence tensor (B, Lmax)
+        :return: ctc loass value
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/frontend_asr.py]: loss

Comment (ShigekiKarita) [reply espnet/nets/pytorch_backend/frontend_asr.py]: I think ASRInterface returns single loss value now. see https://github.com/espnet/espnet/blob/master/espnet/nets/asr_interface.py#L20

File: espnet/nets/pytorch_backend/frontend_asr.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        parser.add_argument('--uttmvn-norm-means', type=strtobool,
+                            default=True, help='')
+        parser.add_argument('--uttmvn-norm-vars', type=strtobool, default=False,
+                            help='')
+        # Feature transform: Fbank
+        parser.add_argument('--fbank-fs', type=int, default=16000,
+                            help='The sample frequency used for '
+                                 'the mel-fbank creation.')
+        parser.add_argument('--n-mels', type=int, default=80,
+                            help='The number of mel-frequency bins.')
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/frontend_asr.py]: Mel

File: espnet/nets/pytorch_backend/frontend_asr.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        parser.add_argument('--wpe-delay', type=int, default=3,
+                            help='')
+        parser.add_argument('--use-dnn-mask-for-wpe', type=strtobool,
+                            default=False,
+                            help='Use DNN to estimate the power spectrogram. '
+                                 'This option is experimental.')
+        # Beamformer related
+        parser.add_argument('--use-beamformer', type=strtobool,
+                            default=True, help='')
+        parser.add_argument('--btype', default='blstmp', type=str,
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/frontend_asr.py]: Can you make these `--bxxx` more expressive like `--masknet-type` or something? because we have already added breaking change in the option `--use-frontend`

File: espnet/nets/pytorch_backend/frontend_asr.py (hunk 7 above)
Comment (ShigekiKarita) [reply espnet/nets/pytorch_backend/frontend_asr.py]: considering `--fbank-xxx` options, `--fbank-dim` is more clear

**Comment by ShigekiKarita** [comment]
@sw005320 Because of the breaking change `--use-frontend` to `--frontend-module`, should we merge this PR into v0.5.0?

**Comment by ShigekiKarita** [comment]
Nice generalization and coverage increase.

This PR looks great except for my small comments above. Some comments might require more discussion. Through this PR, I feel `XXXInterface` is really important to build a nice framework (maybe `ScorerInterface` for beam search is my next work).

**Comment by sw005320** [comment]
> @sw005320 Because of the breaking change --use-frontend to --frontend-module, should we merge this PR into v0.5.0?

I agree.
I'll change it.

Shinji

File: espnet/nets/pytorch_backend/frontend_asr.py (hunk 7 above)
Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/frontend_asr.py]: Mel?

File: espnet/asr/pytorch_backend/asr.py (hunk 4 above)
Comment (kamo-naoyuki) [reply espnet/asr/pytorch_backend/asr.py]: I can't understand out what you mean. This block is inevitable.

**Comment by kamo-naoyuki** [comment]
I mistakenly closed. I reopened at  #890
